### DISCUSSION TITLE
Catch unhandled exceptions within providers

### DISCUSSION THREAD

**Discussion by TreyThomas93** [rrousselGit/riverpod#2015] [Q&A] [answered]
I was wanting to know how to catch  all unhandled exceptions within providers. (not using try/catch)

I have my app wrapped in runZonedGuarded:

```dart
runZonedGuarded(
    () {
      runApp(const ProviderScope(child: GNTMobile()));
    },
    ((error, stack) {
      getIt<LoggerService>().critical(
          error: error, stackTrace: stack, methodName: "runZonedGuarded");
      exit(0);
    }),
  );
```

but it only catches async exceptions. So testing a simple throw in a provider doesn't get caught.

Any ideas on how to do this? Thanks

**Comment by rrousselGit** [comment, marked as answer]
You can combine this with ProviderObserver.
ProviderObserver has to life-cycles for exceptions inside providers

**Reply by TreyThomas93** [reply]
@rrousselGit outstanding! thanks man
